File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int fun(TreeNode root){
        // Agar Root hi null hai; Tree hi gyab hai
        if(root==null){
            return 0;
        }

        // Left Subtree agar nhi hai
        if(root.left==null){
            return 1+fun(root.right);          // Toh Right k node se Height define hogi; Root aur uska Right
        }

        // Agar Right Subtree nhi hai
        if(root.right==null){
            return 1+fun(root.left);           // Toh Tree ki Height Left Subtree se hogi; Root aur uska Left
        }

        // Left Subtree ki Height dekho
        int left=fun(root.left);
        // Right Subtree ki Height dekho
        int right=fun(root.right);

        return 1+Math.min(left,right);
    }

    public int minDepth(TreeNode root) {
        // Recursive Function Call kardo
        return fun(root);
    }
}

File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
int fun(TreeNode*root){
    // Agar root hi null hai ; tree hi gyab hai 
    if(root==nullptr){
        return 0;
    }
    
    // Left subtree agr nhi hai 
    if(root->left==nullptr){
        return 1+fun(root->right);          // toh right k node se ht define hogi ; root or uska immd right
    }
    
    // Agar right subtree nhi hai 
    if(root->right==nullptr){
        return 1+fun(root->left);           // toh tree ka ht. left sub tree se hoga ; root or uska immd left
    }

    // Left subtree ka ht dekho 
    int left=fun(root->left); 
    // Right subtree ka ht dekho 
    int right=fun(root->right);

    return 1+min(left,right);      
}

    int minDepth(TreeNode* root) {
        // Recursive Function call kardo 
        return fun(root);  
    }
};